nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,323 @@
+	} else if len(nodes) == 1 {
+		log.Printf("[INFO] Only one node ip provided, skip joining nodes process")
+		return resourceNsxtManagerClusterRead(d, m)
+	}
+	hostIP, _ := getHostIPFromNsxtClient(m)
+	clusterID, certSha256Thumbprint, err := getClusterInfoFromHostNode(d, m)
+	if err != nil {
+		return err
+	}
+
+	for _, guestNode := range nodes {
+		err := joinNodeToCluster(clusterID, certSha256Thumbprint, guestNode, hostIP, d, m)
+		if err != nil {
+			return err
+		}
+	}
+	d.SetId(clusterID)
+	d.Set("cluster_id", clusterID)
+	return resourceNsxtManagerClusterRead(d, m)
+}
+
+func getHostIPFromNsxtClient(client interface{}) (string, bool) {
+	c := client.(nsxtClients)
+	hostIP := c.Host
+	hostIP = strings.TrimPrefix(hostIP, "https://")
+	ip := net.ParseIP(hostIP)
+	v4 := ip.To4()
+	return hostIP, v4 != nil
+}
+
+func getClusterInfoFromHostNode(d *schema.ResourceData, m interface{}) (string, string, error) {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	clusterConfig, err := client.Get()
+	if err != nil {
+		return "", "", fmt.Errorf("Failed to get cluster info from host node: %s", err)
+	}
+	clusterID := *clusterConfig.ClusterId
+	nodes := clusterConfig.Nodes
+	node := nodes[0]
+	apiListenAddr := node.ApiListenAddr
+	certSha256Thumbprint := *apiListenAddr.CertificateSha256Thumbprint
+	return clusterID, certSha256Thumbprint, nil
+}
+
+func getNewNsxtClient(node NsxClusterNode, d *schema.ResourceData, clients interface{}) (interface{}, error) {
+	// The API of joining a new node to cluster needs to be called with the new node's ip,
+	// to do that we initialize a new nsxtClient with new node's ip as its host
+	username := *node.UserName
+	password := *node.Password
+	host := *node.IPAddress
+	if !strings.HasPrefix(host, "https://") {
+		host = fmt.Sprintf("https://%s", host)
+	}
+	c := clients.(nsxtClients)
+	commonConfig := c.CommonConfig
+	newClients := nsxtClients{
+		CommonConfig: commonConfig,
+	}
+	err := configureNewClient(&newClients, &c, host, username, password)
+	if err != nil {
+		return nil, err
+	}
+	return newClients, nil
+}
+
+func configureNewClient(newClient *nsxtClients, oldClient *nsxtClients, host string, username string, password string) error {
+	newClient.Host = host
+	securityCtx := core.NewSecurityContextImpl()
+	oldSecurityCtx := oldClient.PolicySecurityContext
+	oldSecurityCtxData := oldSecurityCtx.GetAllProperties()
+	_, securityContextNeeded := oldSecurityCtxData[security.AUTHENTICATION_SCHEME_ID]
+	if securityContextNeeded {
+		err := configureSecurityContext(securityCtx, newClient, "", "", "", username, password)
+		if err != nil {
+			return fmt.Errorf("Failed to configure new client with host %s: %s", host, err)
+		}
+	}
+	newClient.PolicyHTTPClient = oldClient.PolicyHTTPClient
+	if securityContextNeeded {
+		newClient.PolicySecurityContext = securityCtx
+	}
+	newClient.PolicyEnforcementPoint = oldClient.PolicyEnforcementPoint
```

Comment:
Those two lines should not be required. Lets avoid using `oldClient` unless absolutely necessary

Reply:
done